drivers/sensor/als31300/als31300.c

```diff
@@ -0,0 +1,681 @@
+		     ((uint32_t)raw_buf[6] << 8) | ((uint32_t)raw_buf[7]);
+
+	/* Extract fields from register 0x28 */
+	temp_msb = (reg28_data & ALS31300_REG28_TEMP_MSB_MASK) >> ALS31300_REG28_TEMP_MSB_SHIFT;
+	z_msb = (reg28_data & ALS31300_REG28_Z_AXIS_MSB_MASK) >> ALS31300_REG28_Z_AXIS_MSB_SHIFT;
+	y_msb = (reg28_data & ALS31300_REG28_Y_AXIS_MSB_MASK) >> ALS31300_REG28_Y_AXIS_MSB_SHIFT;
+	x_msb = (reg28_data & ALS31300_REG28_X_AXIS_MSB_MASK) >> ALS31300_REG28_X_AXIS_MSB_SHIFT;
+
+	/* Extract fields from register 0x29 */
+	temp_lsb = (reg29_data & ALS31300_REG29_TEMP_LSB_MASK) >> ALS31300_REG29_TEMP_LSB_SHIFT;
+	z_lsb = (reg29_data & ALS31300_REG29_Z_AXIS_LSB_MASK) >> ALS31300_REG29_Z_AXIS_LSB_SHIFT;
+	y_lsb = (reg29_data & ALS31300_REG29_Y_AXIS_LSB_MASK) >> ALS31300_REG29_Y_AXIS_LSB_SHIFT;
+	x_lsb = (reg29_data & ALS31300_REG29_X_AXIS_LSB_MASK) >> ALS31300_REG29_X_AXIS_LSB_SHIFT;
+
+	/* Combine MSB and LSB to form 12-bit values */
+	const uint16_t x_raw = (x_msb << 4) | x_lsb;
+	const uint16_t y_raw = (y_msb << 4) | y_lsb;
+	const uint16_t z_raw = (z_msb << 4) | z_lsb;
+
+	temp_raw = (temp_msb << 6) | temp_lsb;
+
+	/* Convert to signed values */
+	x_signed = als31300_convert_12bit_to_signed(x_raw);
+	y_signed = als31300_convert_12bit_to_signed(y_raw);
+	z_signed = als31300_convert_12bit_to_signed(z_raw);
+
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_MAGN_X:
+	case SENSOR_CHAN_MAGN_Y:
+	case SENSOR_CHAN_MAGN_Z:
+	case SENSOR_CHAN_MAGN_XYZ: {
+		struct sensor_three_axis_data *out = data_out;
+
+		out->header.base_timestamp_ns = edata->header.timestamp;
+		out->header.reading_count = 1;
+		out->shift = ALS31300_MAGN_SHIFT;
+
+		/* Convert raw readings to Q31 format */
+		als31300_convert_raw_to_q31_magn(x_signed, &out->readings[0].x);
+		als31300_convert_raw_to_q31_magn(y_signed, &out->readings[0].y);
+		als31300_convert_raw_to_q31_magn(z_signed, &out->readings[0].z);
+		*fit = 1;
+		return 1;
+	}
+	case SENSOR_CHAN_AMBIENT_TEMP: {
+		struct sensor_q31_data *out = data_out;
+
+		out->header.base_timestamp_ns = edata->header.timestamp;
+		out->header.reading_count = 1;
+		out->shift = ALS31300_TEMP_SHIFT;
+
+		als31300_convert_temp_to_q31(temp_raw, &out->readings[0].temperature);
+		*fit = 1;
+		return 1;
+	}
+	default:
+		return -ENOTSUP;
+	}
+}
+
+SENSOR_DECODER_API_DT_DEFINE() = {
+	.get_frame_count = als31300_decoder_get_frame_count,
+	.get_size_info = als31300_decoder_get_size_info,
+	.decode = als31300_decoder_decode,
+};
+
+static int als31300_get_decoder(const struct device *dev, const struct sensor_decoder_api **decoder)
+{
+	ARG_UNUSED(dev);
+	*decoder = &SENSOR_DECODER_NAME();
+	return 0;
+}
+
+#endif /* CONFIG_SENSOR_ASYNC_API */
+
+static const struct sensor_driver_api als31300_api = {
+	.sample_fetch = als31300_sample_fetch,
+	.channel_get = als31300_channel_get,
+#ifdef CONFIG_SENSOR_ASYNC_API
+	.submit = als31300_submit,
+	.get_decoder = als31300_get_decoder,
+#endif
```

Comment:
Should we guard the read/decode-specific code behind an ifdef (like it's done here) or, even better, place it on a separate file that depends on CONFIG_SENSOR_ASYNC_API symbol?